shared-module/displayio/__init__.c

```diff
@@ -117,82 +139,131 @@ static void common_hal_displayio_release_displays_impl(bool keep_primary) {
-            common_hal_is31fl3741_framebuffer_deinit(&display_buses[i].is31fl3741);
+            common_hal_is31fl3741_framebuffer_deinit(DYN_DISPLAY_BUSES_ADR(i, is31fl3741));
         #endif
         #if CIRCUITPY_SHARPDISPLAY
         } else if (bus_type == &sharpdisplay_framebuffer_type) {
-            common_hal_sharpdisplay_framebuffer_deinit(&display_buses[i].sharpdisplay);
+            common_hal_sharpdisplay_framebuffer_deinit(DYN_DISPLAY_BUSES_ADR(i, sharpdisplay));
         #endif
         #if CIRCUITPY_VIDEOCORE
         } else if (bus_type == &videocore_framebuffer_type) {
-            common_hal_videocore_framebuffer_deinit(&display_buses[i].videocore);
+            common_hal_videocore_framebuffer_deinit(DYN_DISPLAY_BUSES_ADR(i, videocore));
         #endif
         #if CIRCUITPY_PICODVI
         } else if (bus_type == &picodvi_framebuffer_type) {
-            common_hal_picodvi_framebuffer_deinit(&display_buses[i].picodvi);
+            common_hal_picodvi_framebuffer_deinit(DYN_DISPLAY_BUSES_ADR(i, picodvi));
         #endif
         }
+        #if CIRCUITPY_OS_GETENV && CIRCUITPY_SET_DISPLAY_LIMIT
+        if (i < CIRCUITPY_DISPLAY_LIMIT) {
+            display_buses[i].bus_base.type = &mp_type_NoneType;
+        } else {
+            display_buses_dyn[i - CIRCUITPY_DISPLAY_LIMIT].bus_base.type = &mp_type_NoneType;
+        }
+        #else
         display_buses[i].bus_base.type = &mp_type_NoneType;
+        #endif
+    }
+    if (!keep_primary) {
+        supervisor_stop_terminal();
     }
-
-    supervisor_stop_terminal();
 }
 
 void common_hal_displayio_release_displays(void) {
     common_hal_displayio_release_displays_impl(false);
 }
 
+void malloc_display_memory(void) {
+    #if CIRCUITPY_OS_GETENV && CIRCUITPY_SET_DISPLAY_LIMIT
+    mp_int_t max_num_displays = CIRCUITPY_DISPLAY_LIMIT;
+    (void)common_hal_os_getenv_int("CIRCUITPY_DISPLAY_LIMIT", &max_num_displays);
+    if (max_num_displays > CIRCUITPY_DISPLAY_LIMIT) {
+        display_buses_dyn = m_new0(primary_display_bus_t, (max_num_displays - CIRCUITPY_DISPLAY_LIMIT));
+        displays_dyn = m_new0(primary_display_t, (max_num_displays - CIRCUITPY_DISPLAY_LIMIT));
+
+        for (uint8_t i = CIRCUITPY_DISPLAY_LIMIT; i < max_num_displays; i++) {
+            display_buses_dyn[i - CIRCUITPY_DISPLAY_LIMIT].bus_base.type = &mp_type_NoneType;
+            displays_dyn[i - CIRCUITPY_DISPLAY_LIMIT].display_base.type = &mp_type_NoneType;
+        }
+        max_allocated_display = max_num_displays;
+    }
+    #endif
+}
+
 void reset_displays(void) {
     // In CircuitPython 10, release secondary displays before doing anything else:
-    // common_hal_displayio_release_displays_impl(true);
+    common_hal_displayio_release_displays_impl(true);
+    #if CIRCUITPY_OS_GETENV && CIRCUITPY_SET_DISPLAY_LIMIT
+    // Does this need to be done or does port_free effectively do this?
+    if (max_allocated_display > CIRCUITPY_DISPLAY_LIMIT) {
```

Comment:
Just posting this to highlight the question of whether this memory cleanup is necessary or would be done by the port_free(_heap) called by stop_mp